discovery/openstack/instance.go

```diff
@@ -0,0 +1,225 @@
+	"github.com/gophercloud/gophercloud/openstack"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/floatingips"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
+	"github.com/gophercloud/gophercloud/pagination"
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+	"golang.org/x/net/context"
+
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/util/strutil"
+)
+
+const (
+	openstackLabelPrefix         = model.MetaLabelPrefix + "openstack_"
+	openstackLabelInstanceID     = openstackLabelPrefix + "instance_id"
+	openstackLabelInstanceName   = openstackLabelPrefix + "instance_name"
+	openstackLabelInstanceStatus = openstackLabelPrefix + "instance_status"
+	openstackLabelInstanceFlavor = openstackLabelPrefix + "instance_flavor"
+	openstackLabelPublicIP       = openstackLabelPrefix + "public_ip"
+	openstackLabelPrivateIP      = openstackLabelPrefix + "private_ip"
+	openstackLabelTagPrefix      = openstackLabelPrefix + "tag_"
+)
+
+// InstanceDiscovery discovers OpenStack instances.
+type InstanceDiscovery struct {
+	authOpts *gophercloud.AuthOptions
+	region   string
+	interval time.Duration
+	logger   log.Logger
+	port     int
+}
+
+// NewInstanceDiscovery returns a new instance discovery.
+func NewInstanceDiscovery(opts *gophercloud.AuthOptions,
+	interval time.Duration, port int, region string, l log.Logger) *InstanceDiscovery {
+	return &InstanceDiscovery{authOpts: opts,
+		region: region, interval: interval, port: port, logger: l}
+}
+
+// Run implements the TargetProvider interface.
+func (i *InstanceDiscovery) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	// Get an initial set right away.
+	tg, err := i.refresh()
+	if err != nil {
+		i.logger.Error(err)
+	} else {
+		select {
+		case ch <- []*config.TargetGroup{tg}:
+		case <-ctx.Done():
+			return
+		}
+	}
+
+	ticker := time.NewTicker(i.interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			tg, err := i.refresh()
+			if err != nil {
+				i.logger.Error(err)
+				continue
+			}
+
+			select {
+			case ch <- []*config.TargetGroup{tg}:
+			case <-ctx.Done():
+				return
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (i *InstanceDiscovery) refresh() (*config.TargetGroup, error) {
+	var err error
+	t0 := time.Now()
+	defer func() {
+		refreshDuration.Observe(time.Since(t0).Seconds())
+		if err != nil {
+			refreshFailuresCount.Inc()
+		}
+	}()
+
+	provider, err := openstack.AuthenticatedClient(*i.authOpts)
+	if err != nil {
+		return nil, fmt.Errorf("could not create OpenStack session: %s", err)
+	}
+	client, err := openstack.NewComputeV2(provider, gophercloud.EndpointOpts{
+		Region: i.region,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("could not create OpenStack compute session: %s", err)
+	}
+
+	opts := servers.ListOpts{}
+	pager := servers.List(client, opts)
+
+	tg := &config.TargetGroup{
+		Source: fmt.Sprintf("OS_" + i.region),
+	}
```

Comment:
Please move the above 6 lines down below the error handling of pagerFIP.

Reply:
Updated.